clang/lib/CIR/Dialect/Transforms/TargetLowering/LowerFunction.cpp

```diff
@@ -32,6 +32,210 @@ using ABIArgInfo = ::cir::ABIArgInfo;
+Value enterStructPointerForCoercedAccess(Value SrcPtr, StructType SrcSTy,
+                                         uint64_t DstSize, LowerFunction &CGF) {
+  // We can't dive into a zero-element struct.
+  if (SrcSTy.getNumElements() == 0)
+    llvm_unreachable("NYI");
+
+  Type FirstElt = SrcSTy.getMembers()[0];
+
+  // If the first elt is at least as large as what we're looking for, or if the
+  // first element is the same size as the whole struct, we can enter it. The
+  // comparison must be made on the store size and not the alloca size. Using
+  // the alloca size may overstate the size of the load.
+  uint64_t FirstEltSize = CGF.LM.getDataLayout().getTypeStoreSize(FirstElt);
+  if (FirstEltSize < DstSize &&
+      FirstEltSize < CGF.LM.getDataLayout().getTypeStoreSize(SrcSTy))
+    return SrcPtr;
+
+  llvm_unreachable("NYI");
+}
+
+/// Create a store to \param Dst from \param Src where the source and
+/// destination may have different types.
+///
+/// This safely handles the case when the src type is larger than the
+/// destination type; the upper bits of the src will be lost.
+void createCoercedStore(Value Src, Value Dst, bool DstIsVolatile,
+                        LowerFunction &CGF) {
+  Type SrcTy = Src.getType();
+  Type DstTy = Dst.getType();
+  if (SrcTy == DstTy) {
+    llvm_unreachable("NYI");
+  }
+
+  // FIXME(cir): We need a better way to handle datalayout queries.
+  assert(isa<IntType>(SrcTy));
+  llvm::TypeSize SrcSize = CGF.LM.getDataLayout().getTypeAllocSize(SrcTy);
+
+  if (StructType DstSTy = dyn_cast<StructType>(DstTy)) {
+    Dst = enterStructPointerForCoercedAccess(Dst, DstSTy,
+                                             SrcSize.getFixedValue(), CGF);
+    assert(isa<PointerType>(Dst.getType()));
+    DstTy = cast<PointerType>(Dst.getType()).getPointee();
+  }
+
+  PointerType SrcPtrTy = dyn_cast<PointerType>(SrcTy);
+  PointerType DstPtrTy = dyn_cast<PointerType>(DstTy);
+  // TODO(cir): Implement address space.
+  if (SrcPtrTy && DstPtrTy && !::cir::MissingFeatures::addressSpace()) {
+    llvm_unreachable("NYI");
+  }
+
+  // If the source and destination are integer or pointer types, just do an
+  // extension or truncation to the desired type.
+  if ((isa<IntegerType>(SrcTy) || isa<PointerType>(SrcTy)) &&
+      (isa<IntegerType>(DstTy) || isa<PointerType>(DstTy))) {
+    llvm_unreachable("NYI");
+  }
+
+  llvm::TypeSize DstSize = CGF.LM.getDataLayout().getTypeAllocSize(DstTy);
+
+  // If store is legal, just bitcast the src pointer.
+  assert(!::cir::MissingFeatures::vectorType());
+  if (SrcSize.getFixedValue() <= DstSize.getFixedValue()) {
+    // Dst = Dst.withElementType(SrcTy);
+    CGF.buildAggregateStore(Src, Dst, DstIsVolatile);
+  } else {
+    llvm_unreachable("NYI");
+  }
+}
+
+// FIXME(cir): Create a custom rewriter class to abstract this away.
+Value createBitcast(Value Src, Type Ty, LowerFunction &LF) {
+  return LF.getRewriter().create<CastOp>(Src.getLoc(), Ty, CastKind::bitcast,
```

Comment:
Is this where the bitcast between struct comes from?

Reply:
In the literal way, yes, this is the line that creates it. Semantically, we need the bitcast to circumvent MLIR's typechecking and properly abide to the calling convention by re-interpreting structs as integers.